src/nodes/NodeConnectionManager.ts

```diff
@@ -0,0 +1,714 @@
+import type ReverseProxy from '../network/ReverseProxy';
+import type ForwardProxy from '../network/ForwardProxy';
+import type { Host, Hostname, Port } from '../network/types';
+import type { ResourceAcquire } from '../utils';
+import type { Timer } from '../types';
+import type NodeGraph from './NodeGraph';
+import type { NodeId, NodeAddress, NodeData, SeedNodes } from './types';
+import Logger from '@matrixai/logger';
+import { StartStop, ready } from '@matrixai/async-init/dist/StartStop';
+import { IdInternal } from '@matrixai/id';
+import { status } from '@matrixai/async-init';
+import NodeConnection from './NodeConnection';
+import * as nodesUtils from './utils';
+import * as nodesErrors from './errors';
+import GRPCClientAgent from '../agent/GRPCClientAgent';
+import * as validationUtils from '../validation/utils';
+import * as networkUtils from '../network/utils';
+import * as agentErrors from '../agent/errors';
+import * as grpcErrors from '../grpc/errors';
+import * as nodesPB from '../proto/js/polykey/v1/nodes/nodes_pb';
+import { RWLock, withF } from '../utils';
+
+type ConnectionAndLock = {
+  connection?: NodeConnection<GRPCClientAgent>;
+  timer?: NodeJS.Timer;
+  lock: RWLock;
+};
+
+/**
+ * Data structure to store all NodeConnections. If a connection to a node n does
+ * not exist, no entry for n will exist in the map. Alternatively, if a
+ * connection is currently being instantiated by some thread, an entry will
+ * exist in the map, but only with the lock (no connection object). Once a
+ * connection is instantiated, the entry in the map is updated to include the
+ * connection object.
+ */
+type NodeConnectionMap = Map<string, ConnectionAndLock>;
+
+/**
+ * This interface is provided to each `NodeConnection` in the connection map
+ * to provide necessary features. The main one being `killSelf` to allow the
+ * `NodeConnection` to tell the `NodeConnectionManager` to destroy the
+ * connection thus removing it from  the connection map.
+ */
+export interface NodeConnectionManagerInterface {
+  withConnF: NodeConnectionManager['withConnF'];
+  getSeedNodes: NodeConnectionManager['getSeedNodes'];
+  sendHolePunchMessage: NodeConnectionManager['sendHolePunchMessage'];
+  destroyCallback(): Promise<void>;
+}
+
+interface NodeConnectionManager extends StartStop {}
+@StartStop()
+class NodeConnectionManager {
+  /**
+   * Time used to estalish `NodeConnection`
+   */
+  public readonly connConnectTime: number;
+
+  /**
+   * Time to live for `NodeConnection`
+   */
+  public readonly connTimeoutTime: number;
+  /**
+   * Alpha constant for kademlia
+   * The number of closest nodes to contact initially
+   */
+  public readonly initialClosestNodes: number;
+
+  protected logger: Logger;
+  protected nodeGraph: NodeGraph;
+  protected keyManager: KeyManager;
+  protected fwdProxy: ForwardProxy;
+  protected revProxy: ReverseProxy;
+  protected seedNodes: SeedNodes;
+  protected connections: NodeConnectionMap = new Map();
```

Comment:
Move `NodeConnectionMap` type to be embedded directly here. And then you can move the above docblock comment here too.

Reply:
This is done.